### ISSUE TITLE
dtor of return value isn't run if dtor of a scoped variable throws and is caught in the same function

### ISSUE THREAD

**Issue by kelbon** [llvm/llvm-project#73024]
https://godbolt.org/z/nYnvv3EEf

Code without address sanitizer to avoid false positive:

https://godbolt.org/z/qbxMG659h

```cpp

#include <vector>

inline bool g = false;

struct X {
    int i = 0;

    ~X() {
        if (i == 1) g = true;
    }
};
struct Y {
    ~Y() noexcept(false) { throw 0; }
};
X f() {
    try {
        Y y;
        return X{1};
    } catch (...) {
    }
    return X{};
}

int main() {
    f();
    if (!g)
      return 4;
}

```

**Comment by dwblaikie** [comment]
Can't say I understand the non-asan failure. It looks correct, but I'm probably misunderstanding... 

In https://godbolt.org/z/qxTjva39r - the Y dtor runs, so `X{1}` does /not/ execute, but `X{}` does execute, so `g` would still be false, and main would return 4, right? That seems correct?

**Comment by kelbon** [comment]
> Can't say I understand the non-asan failure. It looks correct, but I'm probably misunderstanding...
> 
> In https://godbolt.org/z/qxTjva39r - the Y dtor runs, so `X{1}` does /not/ execute, but `X{}` does execute, so `g` would still be false, and main would return 4, right? That seems correct?

No, its not correct, value was constructed, but not destroyed, basic RAII broken

https://godbolt.org/z/sqfeT7jPh

**Comment by dwblaikie** [comment]
Yeah, fair enough, that last one seems unambiguously buggy to me.
